linalg/hypre.hpp

```diff
@@ -989,9 +1034,83 @@ class HypreBoomerAMG : public HypreSolver
        As with SetSystemsOptions(), this solver assumes Ordering::byVDIM. */
    void SetElasticityOptions(ParFiniteElementSpace *fespace);
 
+#if MFEM_HYPRE_VERSION >= 21800
+   /* distance parameter takes on values {1,2,15} for lAIR, meaning R is built using
+   distance 1 neighbors, distance two neighbors, or distance two on processor and
+   distance 1 off processor (i.e., distance 1.5 --> 15).        */
+   void SetLAIROptions(int distance=15,  std::string prerelax="",
+                       std::string postrelax="FFC", double strength_tol=0.1,
+                       double strength_tolR=0.01, double filter_tolR=0.0,
+                       int interp_type=100, int relax_type=3, double filterA_tol=0.0,
+                       int splitting=6, int blksize=0, int Sabs=0);
+
+   void SetNAIROptions(int neumann_degree=2,  std::string prerelax="A",
+                       std::string postrelax="F", double strength_tol=0.1,
+                       double strength_tolR=0.01, double filter_tolR=0.0,
+                       int interp_type=100, int relax_type=10, double filterA_tol=0.0,
+                       int splitting=6, int blksize=0, int Sabs=0);
+
+   void SetStrengthThreshR(double strengthR)
```

Comment:
As someone not officially reviewing this, I'm reluctant to jump in, but I think adding all these `Set` methods for options is a mistake. As I understand it, the MFEM interface to Hypre is intended to only expose a few most commonly used options for simplicity, and if you want finer control you need to consult the Hypre documentation and use the `virtual operator HYPRE_Solver() const` method to use the MFEM object as a Hypre object, ie, you can do
```c++
mfem::HypreBoomerAMG amg(A);
HYPRE_BoomerAMGSetStrongThresholdR(amg, 0.1);
HYPRE_BoomerAMGSetIsTriangular(amg, 1);
```
without adding all these methods to the classes.

Reply:
I'm not on the development team but do disagree here. I don't see any particular usefulness of having the backend boomerang class be "simple," by which you mean just not having direct access to pretty standard AMG options. All the functions are one-liners in the header file - what is the advantage of removing these? Something like HYPRE_BoomerAMGSetIsTriangular is specific enough that I don't feel strongly about it being there, but think there should absolutely be functions for coarsening, interpolation, strength tolerance, etc. @barker29 you are well aware how sensitive AMG can be to these choice; not providing obvious functionality to set them does not make sense to me.

Reply:
The functionality *is* provided, it just goes through the Hypre interface rather than having an MFEM wrapper. I just don't see an advantage in adding a bunch of methods to `HypreBoomerAMG` so that users can spell things with a C++ convention rather than a C convention. But I'll shut up now and let those actually involved in this PR hash it out.

I should add, to be clear, I like the idea of having the AIR solver integrated in MFEM and in our advection examples, so I would actually be very happy to see this PR merged (regardless of our decision on this minor interface issue).

Reply:
I don't have an opinion on this, but I suspect others (@tzanio? @v-dobrev?) might.

Reply:
Thanks for bringing this up @barker29. I think it is actually an important topic and we should have a discussion about it. I'll post my thoughts below, but maybe a telecon will be more efficient way to discuss this?

To me, the goal of the MFEM interface for hypre is to make it easy for the user to construct a solver that in most cases will just work for their problem.

This targets people that just want to solve a system and do not want to learn or explore all the AMG options. If that's what they want to do, then they should use the hypre interfaces directly, there is no added value to use the MFEM ones.

For the first class of users thought, there is added value, which is the "curated" set of options. For example:

- If you want to solve an H1 problem, (implicitly)  call `HypreBoomerAMG::SetDefaultOptions()`;
- If the problem is a block system, call `HypreBoomerAMG::SetSystemsOptions(int dim)` instead;
- If the problem is linear Elasticity, call `HypreBoomerAMG::SetElasticityOptions(ParFiniteElementSpace *fespace)`.

We should of course try to update these defaults to be as good as generally possible and add new sets of options, e.g. `SetConservativeOptions`, `SetAdvectionOptions`, etc., but I think it is better to group the options depending on the goal that the user wants to achieve.

If instead the user knows about `HYPRE_BoomerAMGSetStrongThreshold` and all the other AMG options and wants to experiment with them, then they should talk to the hypre team, and adding additional level of redirection will just stand in the way IMHO. I also doubt that we will be able to keep up with all the new AMG options :)

In this particular PR, I will suggest that you consider using a different name for SetLAIROptions and SetNAIROptions and make the arguments internal. Maybe that's not critical, but what does the user trying to achieve when they call these functions? How do the two functions differ. A good name will ideally answer these questions without the need to examine the code (and know about the meaning of hypre options).

Just my 2¢

Reply:
I don't see a big drawback to allow both:
```c++
mfem::HypreBoomerAMG amg(A);
HYPRE_BoomerAMGSetStrongThresholdR(amg, 0.1);
HYPRE_BoomerAMGSetIsTriangular(amg, 1);
```
and
```c++
mfem::HypreBoomerAMG amg(A);
amg.SetStrongThresholdR(0.1);
amg.SetIsTriangular(1);
```
I don't think having these "expert" options will confuse users -- we can put them in an "Expert options" section in the code (and in doxygen). We can also add "This is an expert option" doxygen comments to all such methods.

On the other hand, I agree with @tzanio that simple, user-friendly options tailored to different use cases will be most helpful.

Reply:
I am willing to be convinced :).

But at a minimum the names should be the same, i.e. `amg.SetStrongThresholdR()` vs. the current `SetStrengthThreshR()`...

Reply:
I agree, there should be 1-to-1 correspondence with the hypre function for such wrappers. For example, I'd prefer to avoid "expert" methods that do more than 1 hypre call, e.g. as in `SetRelaxCycle`, or `SetNodal`.